csrc/host_ir/jit.cpp

```diff
@@ -259,65 +363,143 @@ HostIrJit::HostIrJit(hir::HostIrContainer* container, int num_threads)
   llvm::orc::SymbolMap symbolMap;
-  symbolMap[mangler(kHostIrJitAtEmptyStridedCudaFuncName)] =
-      llvm::orc::ExecutorSymbolDef(
-          at_empty_strided_cuda_addr, llvm::JITSymbolFlags::Exported);
-
+  registerExternalFunction(
+      extract_tensor_size_func_ptr, symbolMap, mangler, kTensorSizeFuncName);
+  registerExternalFunction(
+      allocate_tensor_func_ptr, symbolMap, mangler, kAllocateTensorFuncName);
+  registerExternalFunction(
+      set_tensor_func_ptr, symbolMap, mangler, kSetTensorFuncName);
   throwIfError(dest_dynamic_lib.define(llvm::orc::absoluteSymbols(symbolMap)));
-
   // Compile the module
-  compile(container, pimpl_->jit.get(), pimpl_->allocate_funcs_);
+  compile(pimpl_.get());
 }
 
-at::Tensor HostIrJit::allocate(
-    const kir::Allocate* allocate,
-    const std::vector<int64_t>& input_sizes,
-    const std::vector<int64_t>& input_strides) {
-  FUSER_PERF_SCOPE("HostIrJit::allocate");
-  auto allocate_func_iter = pimpl_->allocate_funcs_.find(allocate);
-  NVF_ERROR(
-      allocate_func_iter != pimpl_->allocate_funcs_.end(),
-      "allocate function not found for ",
-      allocate);
-  const auto& func_ptr = allocate_func_iter->second;
-
-  at::Tensor tensor;
-  func_ptr(
-      input_sizes.data(),
-      input_sizes.size(),
-      input_strides.data(),
-      input_strides.size(),
-      tensor);
-
-  return tensor;
+KernelArgumentHolder HostIrJit::runWithInputs(
+    const KernelArgumentHolder& args) {
+  FUSER_PERF_SCOPE("HostIrJit::runWithInputs");
+  // Bind cache id to llvm global variable or align with main function inputs
+  NVF_ERROR(args.getCacheId().has_value(), "Cache ID is not set");
+  NVF_ERROR_EQ(std::ssize(pimpl_->container->inputs()), args.size());
+
+  std::vector<const void*> input_aten_tensors;
+  // Bind the inputs to the tensor map
+  for (auto&& [in_val, arg] : zip(pimpl_->container->inputs(), args)) {
+    if (arg.is<at::Tensor>()) {
+      input_aten_tensors.push_back(&arg.as<at::Tensor>());
+    } else {
+      // TODO: handle other primitive types so we can just align them to input
+      // of main function
+      NVF_THROW("Unsupported argument type: ", arg);
+    }
+  }
+
+  // Run the main function
+  void** output_aten_tensors = pimpl_->main_func(input_aten_tensors.data());
+
+  // Collect the outputs
+  KernelArgumentHolder outputs;
+  for (size_t i = 0; i < pimpl_->container->outputs().size(); ++i) {
+    auto* output = pimpl_->container->outputs()[i];
+    if (output->isA<TensorView>()) {
+      // Cast void* to at::Tensor* first, then dereference
+      at::Tensor* tensor_ptr = static_cast<at::Tensor*>(output_aten_tensors[i]);
+      outputs.push(*tensor_ptr);
+      // Clean up the individual tensor object (not the array)
+      delete tensor_ptr;
+    } else {
+      NVF_THROW("Unsupported output type: ", output);
+    }
+  }
+  // Note: output_aten_tensors points to a global array managed by JIT, don't
```

Comment:
> a global array managed by JIT

I'm not sure about the implications of that? Could there be a race condition where multiple host IR executions want to write to it? 

Is it simpler for **this function** to allocate `output_aten_tensors` (just as how it makes `input_aten_tensors`) and pass it into main_func? This way, main_func takes two `Tensor**`s, one for input and one for output.